File: iterator.go
package k2tree

// iterFrame holds the cursor state at one T level.
type iterFrame struct {
	sublayeroff int // which block at this level
	val         int // column value at this level (bits below this level are 0)
	offInRun    int // current bit within the row's block: x*kPerLayer + y
	runCount    int // Count(levelStart, levelStart+sublayeroff*bitsPerLayer+offInRun)
	baseOff     int // x * kPerLayer — row's fixed component at this T level
}

// Iterator iterates over edges in a K2Tree.
type Iterator struct {
	tree   *K2Tree
	rowcol int
	isRow  bool
	offset int // last returned column (-1 = not yet started)
	done   bool

	// tstack[i] is the frame for T level (i+1).
	// tstack[0] = level 1 (lowest T level, parent of leaf).
	// tstack[tree.levels-1] = top T level.
	tstack     []iterFrame
	leafSub    int // sublayeroff for the leaf
	leafCol    int // current column in the leaf
	leafBitoff int // offsetL(rowcol, it.offset) — bitoff of the last returned column
}

func newRowIterator(tree *K2Tree, row int) *Iterator {
	it := &Iterator{
		tree:   tree,
		rowcol: row,
		isRow:  true,
		offset: -1,
		done:   tree.levels == 0,
	}
	if tree.levels > 0 {
		it.tstack = make([]iterFrame, tree.levels)
		for i := 0; i < tree.levels; i++ {
			// baseOff = x * kPerLayer, the row's fixed component at T level i+1.
			it.tstack[i].baseOff = tree.offsetTForLayer(row, 0, i+1)
		}
	}
	return it
}

func newColumnIterator(tree *K2Tree, col int) *Iterator {
	it := &Iterator{
		tree:   tree,
		rowcol: col,
		isRow:  false,
		offset: -1,
		done:   tree.levels == 0,
	}
	if tree.levels > 0 {
		it.tstack = make([]iterFrame, tree.levels)
		for i := 0; i < tree.levels; i++ {
			// baseOff = y * 1 — column's fixed component at T level i+1.
			it.tstack[i].baseOff = tree.offsetTForLayer(0, col, i+1)
		}
	}
	return it
}

func (it *Iterator) Next() bool {
	if it.done {
		return false
	}
	if it.offset == -1 {
		// First call: descend from the top.
		if !it.descendInto(it.tree.levels-1, 0, 0) {
			it.done = true
			return false
		}
		return true
	}

	// Try to advance within the current leaf block.
	// If the next value's bitoff increases, it might be in the same block.
	// advanceLeaf handles both same-block and cross-block cases.
	nextCol := it.offset + 1
	nextBitoff := it.offsetLForVal(nextCol)
	if nextBitoff > it.leafBitoff {
		it.leafCol = nextCol
		if it.advanceLeaf() {
			return true
		}
	}

	// Leaf block exhausted (or next col is in a new block). Bubble up.
	for i := 0; i < it.tree.levels; i++ {
		frame := &it.tstack[i]
		level := i + 1
		levelStart := it.tree.levelOffsets[level]

		// The bit at frame.offInRun was SET (we descended from it).
		// Advance past it: the set bit contributes +1 to runCount.
		frame.runCount++
		newVal := it.tree.incrementNForLevel(frame.val, 1, level)
		newOffInRun := it.offsetTForVal(newVal, level)
		if newOffInRun < frame.offInRun {
			// Row's block at this level is exhausted; bubble up further.
			continue
		}
		frame.val = newVal
		frame.offInRun = newOffInRun

		// Scan-and-descend loop: handles dead ends (subtrees with bits for
		// other rows but none for it.rowcol) by advancing within this level.
		for {
			if !it.scanFrame(i) {
				break // this level's block is exhausted; bubble up
			}

			// Recalculate runCount after scanFrame (it may have skipped set bits).
			bitoff := levelStart + frame.sublayeroff*it.tree.tk.bitsPerLayer + frame.offInRun
			frame.runCount = it.tree.tbits.Count(levelStart, bitoff)

			if it.descendInto(i-1, frame.runCount, frame.val) {
				return true
			}

			// descendInto returned false: no result for this row in that subtree.
			// Advance past the current set bit and retry at this level.
			frame.runCount++
			newVal2 := it.tree.incrementNForLevel(frame.val, 1, level)
			newOffInRun2 := it.offsetTForVal(newVal2, level)
			if newOffInRun2 < frame.offInRun {
				break // this level's block is now exhausted
			}
			frame.val = newVal2
			frame.offInRun = newOffInRun2
		}
	}

	it.done = true
	return false
}

// offsetTForVal returns the offset of (rowcol, val) or (val, rowcol) in a T block.
func (it *Iterator) offsetTForVal(val, level int) int {
	if it.isRow {
		return it.tree.offsetTForLayer(it.rowcol, val, level)
	}
	return it.tree.offsetTForLayer(val, it.rowcol, level)
}

// offsetLForVal returns the leaf offset for the current (rowcol, val) pair.
func (it *Iterator) offsetLForVal(v int) int {
	if it.isRow {
		return it.tree.offsetL(it.rowcol, v)
	}
	return it.tree.offsetL(v, it.rowcol)
}

// scanFrame scans tstack[i] from frame.offInRun for the next set bit.
// On each unset bit it advances frame.val and frame.offInRun; runCount is
// NOT modified (unset bits don't change it — the caller increments runCount
// when moving past a set bit).
// Returns true when a set bit is found at the updated frame.offInRun.
func (it *Iterator) scanFrame(i int) bool {
	frame := &it.tstack[i]
	level := i + 1
	levelStart := it.tree.levelOffsets[level]
	for {
		bitoff := levelStart + frame.sublayeroff*it.tree.tk.bitsPerLayer + frame.offInRun
		if it.tree.tbits.Get(bitoff) {
			return true
		}
		// Unset: advance val/offInRun; runCount stays the same.
		newVal := it.tree.incrementNForLevel(frame.val, 1, level)
		newOffInRun := it.offsetTForVal(newVal, level)
		if newOffInRun < frame.offInRun {
			// Wrapped: block exhausted.
			frame.val = newVal
			frame.offInRun = newOffInRun
			return false
		}
		frame.val = newVal
		frame.offInRun = newOffInRun
	}
}

// descendInto initialises tstack[startIdx] down to tstack[0], then the leaf.
// sublayeroff is the child-block index for the block at tstack[startIdx].
// val is the starting column (bits for levels ≤ startIdx are 0).
// Returns true if a value was found and stored in it.offset.
//
// Dead ends — where a T-level bit is set by another row sharing the same
// x-path but diverging at a lower level — are handled by retrying within
// this level before returning false.
func (it *Iterator) descendInto(startIdx int, sublayeroff int, val int) bool {
	if startIdx < 0 {
		it.leafSub = sublayeroff
		it.leafCol = val
		return it.advanceLeaf()
	}

	frame := &it.tstack[startIdx]
	level := startIdx + 1
	levelStart := it.tree.levelOffsets[level]

	frame.sublayeroff = sublayeroff
	frame.val = val
	frame.offInRun = frame.baseOff

	// Find the first set bit; one Count call to initialise runCount.
	if !it.scanFrame(startIdx) {
		return false
	}
	bitoff := levelStart + sublayeroff*it.tree.tk.bitsPerLayer + frame.offInRun
	frame.runCount = it.tree.tbits.Count(levelStart, bitoff)

	for {
		if it.descendInto(startIdx-1, frame.runCount, frame.val) {
			return true
		}

		// Sub-level returned false (dead end or genuinely empty for this row).
		// Advance past the current set bit — it contributes +1 to runCount —
		// then scan for the next set bit.
		frame.runCount++
		newVal := it.tree.incrementNForLevel(frame.val, 1, level)
		newOffInRun := it.offsetTForVal(newVal, level)
		if newOffInRun < frame.offInRun {
			return false // this level's block exhausted
		}
		frame.val = newVal
		frame.offInRun = newOffInRun

		if !it.scanFrame(startIdx) {
			return false // this level's block exhausted after scanning
		}
		// Recalculate runCount after scanFrame (it may have skipped set bits).
		bitoff2 := levelStart + frame.sublayeroff*it.tree.tk.bitsPerLayer + frame.offInRun
		frame.runCount = it.tree.tbits.Count(levelStart, bitoff2)
	}
}

// advanceLeaf scans forward from it.leafCol within the current leaf block
// (identified by it.leafSub). Updates it.leafCol, it.leafBitoff, and
// it.offset on success; returns false when the leaf block is exhausted.
func (it *Iterator) advanceLeaf() bool {
	leafoffset := it.leafSub * it.tree.lk.bitsPerLayer
	v := it.leafCol
	bitoff := it.offsetLForVal(v)
	for {
		if it.tree.lbits.Get(leafoffset + bitoff) {
			it.leafCol = v
			it.leafBitoff = bitoff
			it.offset = v
			return true
		}
		v++
		newbitoff := it.offsetLForVal(v)
		if newbitoff < bitoff {
			it.leafCol = v
			return false // leaf block exhausted
		}
		bitoff = newbitoff
	}
}

func (it *Iterator) Value() int {
	return it.offset
}

func (it *Iterator) ExtractAll() []int {
	var out []int
	for it.Next() {
		out = append(out, it.Value())
	}
	return out
}

File: rust/src/k2tree/iterator.rs
use super::K2Tree;
use crate::bitarray::BitArray;

struct IterFrame {
    sublayeroff: usize,
    val: usize,
    off_in_run: usize,
    run_count: usize,
    base_off: usize,
}

/// K2TreeIterator iterates over edges in a K2Tree.
pub struct K2TreeIterator<'a, T: BitArray> {
    tree: &'a K2Tree<T>,
    rowcol: usize,
    is_row: bool,
    offset: isize,
    done: bool,
    tstack: Vec<IterFrame>,
    leaf_sub: usize,
    leaf_col: usize,
    leaf_bitoff: usize,
}

impl<'a, T: BitArray> K2TreeIterator<'a, T> {
    /// Creates a new row iterator (outgoing edges from node row).
    pub fn new_row(tree: &'a K2Tree<T>, row: usize) -> Self {
        let done = tree.levels == 0;
        let tstack = if tree.levels > 0 {
            (0..tree.levels)
                .map(|i| IterFrame {
                    sublayeroff: 0,
                    val: 0,
                    off_in_run: 0,
                    run_count: 0,
                    base_off: tree.offset_t_for_layer(row, 0, i + 1),
                })
                .collect()
        } else {
            Vec::new()
        };
        K2TreeIterator {
            tree,
            rowcol: row,
            is_row: true,
            offset: -1,
            done,
            tstack,
            leaf_sub: 0,
            leaf_col: 0,
            leaf_bitoff: 0,
        }
    }

    /// Creates a new column iterator (incoming edges to node col).
    pub fn new_column(tree: &'a K2Tree<T>, col: usize) -> Self {
        let done = tree.levels == 0;
        let tstack = if tree.levels > 0 {
            (0..tree.levels)
                .map(|i| IterFrame {
                    sublayeroff: 0,
                    val: 0,
                    off_in_run: 0,
                    run_count: 0,
                    base_off: tree.offset_t_for_layer(0, col, i + 1),
                })
                .collect()
        } else {
            Vec::new()
        };
        K2TreeIterator {
            tree,
            rowcol: col,
            is_row: false,
            offset: -1,
            done,
            tstack,
            leaf_sub: 0,
            leaf_col: 0,
            leaf_bitoff: 0,
        }
    }

    /// Like `offset_t_for_layer` but swaps arguments for column iteration.
    fn offset_t_for_val(&self, val: usize, level: usize) -> usize {
        if self.is_row {
            self.tree.offset_t_for_layer(self.rowcol, val, level)
        } else {
            self.tree.offset_t_for_layer(val, self.rowcol, level)
        }
    }

    /// Like `offset_l` but swaps arguments for column iteration.
    fn offset_l_for_val(&self, val: usize) -> usize {
        if self.is_row {
            self.tree.offset_l(self.rowcol, val)
        } else {
            self.tree.offset_l(val, self.rowcol)
        }
    }

    /// Advances the iterator to the next edge.
    /// Returns true if there is a next value, false if iteration is complete.
    pub fn next_edge(&mut self) -> bool {
        if self.done {
            return false;
        }
        if self.offset == -1 {
            if !self.descend_into(self.tree.levels as isize - 1, 0, 0) {
                self.done = true;
                return false;
            }
            return true;
        }

        // Try to advance within the current leaf block.
        let next_col = (self.offset + 1) as usize;
        let next_bitoff = self.offset_l_for_val(next_col);
        if next_bitoff > self.leaf_bitoff {
            self.leaf_col = next_col;
            if self.advance_leaf() {
                return true;
            }
        }

        // Leaf block exhausted. Bubble up through T levels.
        for i in 0..self.tree.levels {
            let level = i + 1;
            let level_start = self.tree.level_offsets[level];

            self.tstack[i].run_count += 1;
            let new_val = self.tree.increment_n_for_level(self.tstack[i].val, 1, level);
            let new_off_in_run = self.offset_t_for_val(new_val, level);
            if new_off_in_run < self.tstack[i].off_in_run {
                continue;
            }
            self.tstack[i].val = new_val;
            self.tstack[i].off_in_run = new_off_in_run;

            loop {
                if !self.scan_frame(i) {
                    break;
                }

                // Recalculate run_count after scan_frame (it may have skipped set bits).
                let bitoff = level_start
                    + self.tstack[i].sublayeroff * self.tree.tk.bits_per_layer
                    + self.tstack[i].off_in_run;
                self.tstack[i].run_count = self.tree.tbits.count(level_start, bitoff);

                let run_count = self.tstack[i].run_count;
                let val = self.tstack[i].val;
                if self.descend_into(i as isize - 1, run_count, val) {
                    return true;
                }

                self.tstack[i].run_count += 1;
                let new_val2 = self.tree.increment_n_for_level(self.tstack[i].val, 1, level);
                let new_off2 = self.offset_t_for_val(new_val2, level);
                if new_off2 < self.tstack[i].off_in_run {
                    break;
                }
                self.tstack[i].val = new_val2;
                self.tstack[i].off_in_run = new_off2;
            }
        }

        self.done = true;
        false
    }

    /// Returns the current value (the node index of the current edge).
    pub fn value(&self) -> usize {
        self.offset as usize
    }

    /// Extracts all remaining values into a Vec.
    pub fn extract_all(&mut self) -> Vec<usize> {
        let mut out = Vec::new();
        while self.next_edge() {
            out.push(self.value());
        }
        out
    }

    // scan_frame scans tstack[i] from off_in_run for the next set bit.
    // Advances val and off_in_run on each unset bit; run_count is NOT modified.
    // Returns true when a set bit is found.
    fn scan_frame(&mut self, i: usize) -> bool {
        let level = i + 1;
        let level_start = self.tree.level_offsets[level];
        loop {
            let bitoff = level_start
                + self.tstack[i].sublayeroff * self.tree.tk.bits_per_layer
                + self.tstack[i].off_in_run;
            if self.tree.tbits.get(bitoff) {
                return true;
            }
            let new_val = self.tree.increment_n_for_level(self.tstack[i].val, 1, level);
            let new_off = self.offset_t_for_val(new_val, level);
            if new_off < self.tstack[i].off_in_run {
                self.tstack[i].val = new_val;
                self.tstack[i].off_in_run = new_off;
                return false;
            }
            self.tstack[i].val = new_val;
            self.tstack[i].off_in_run = new_off;
        }
    }

    // descend_into initialises tstack[start_idx] down to tstack[0], then the leaf.
    // Returns true if a value was found and stored in self.offset.
    fn descend_into(&mut self, start_idx: isize, sublayeroff: usize, val: usize) -> bool {
        if start_idx < 0 {
            self.leaf_sub = sublayeroff;
            self.leaf_col = val;
            return self.advance_leaf();
        }

        let i = start_idx as usize;
        let level = i + 1;
        let level_start = self.tree.level_offsets[level];

        self.tstack[i].sublayeroff = sublayeroff;
        self.tstack[i].val = val;
        self.tstack[i].off_in_run = self.tstack[i].base_off;

        if !self.scan_frame(i) {
            return false;
        }

        let bitoff = level_start
            + sublayeroff * self.tree.tk.bits_per_layer
            + self.tstack[i].off_in_run;
        self.tstack[i].run_count = self.tree.tbits.count(level_start, bitoff);

        loop {
            let run_count = self.tstack[i].run_count;
            let val = self.tstack[i].val;
            if self.descend_into(start_idx - 1, run_count, val) {
                return true;
            }

            self.tstack[i].run_count += 1;
            let new_val = self.tree.increment_n_for_level(self.tstack[i].val, 1, level);
            let new_off = self.offset_t_for_val(new_val, level);
            if new_off < self.tstack[i].off_in_run {
                return false;
            }
            self.tstack[i].val = new_val;
            self.tstack[i].off_in_run = new_off;

            if !self.scan_frame(i) {
                return false;
            }
            // Recalculate run_count after scan_frame (it may have skipped set bits).
            let bitoff2 = level_start
                + self.tstack[i].sublayeroff * self.tree.tk.bits_per_layer
                + self.tstack[i].off_in_run;
            self.tstack[i].run_count = self.tree.tbits.count(level_start, bitoff2);
        }
    }

    // advance_leaf scans forward from leaf_col within the current leaf block.
    // Updates leaf_col, leaf_bitoff, and offset on success.
    fn advance_leaf(&mut self) -> bool {
        let leafoffset = self.leaf_sub * self.tree.lk.bits_per_layer;
        let mut val = self.leaf_col;
        let mut bitoff = self.offset_l_for_val(val);
        loop {
            if self.tree.lbits.get(leafoffset + bitoff) {
                self.leaf_col = val;
                self.leaf_bitoff = bitoff;
                self.offset = val as isize;
                return true;
            }
            val += 1;
            let new_bitoff = self.offset_l_for_val(val);
            if new_bitoff < bitoff {
                self.leaf_col = val;
                return false;
            }
            bitoff = new_bitoff;
        }
    }
}
